File: Source/Json/CPP/Json.cpp
#include "../../Assert/CPP/Assert.hpp"
#include "../../InternalString/CPP/InternalString.hpp"
#include "../../ExternalString/CPP/ExternalString.hpp"
#include "../../Blob/CPP/Blob.hpp"
#include "Json.hpp"

#include <cstdio>

namespace NumbatLogic
{
	Json::Json()
	{
		m_pOwnedJson = NULL;
		m_pJson = NULL;
		m_pNext = NULL;
		m_pPrevious = NULL;
		m_pChild = NULL;
	}

	Json::Json(cJSON* pJson, Json* pPrevious)
	{
		Assert::Plz(pJson);

		m_pOwnedJson = NULL;
		m_pJson = pJson;
		m_pPrevious = pPrevious;
		m_pNext = NULL;
		m_pChild = NULL;

		RecursiveCreate();
	}

	Json::~Json()
	{
		Cleanup();
	}

	bool Json::LoadFromExternalString(const char* szIn)
	{
		Cleanup();

		if (szIn != 0)
		{
			m_pOwnedJson = cJSON_Parse(szIn);
			m_pJson = m_pOwnedJson;

			if (RecursiveCreate())
				return true;
		}

		return false;
	}

	bool Json::LoadFromPath(const char* sxPath)
	{
		bool bSuccess = false;
		gsBlob* pBlob = new gsBlob();
		
		if (!pBlob->Load(sxPath))
			goto Cleanup;
		pBlob->SetOffset(pBlob->GetSize());
		pBlob->PackUint8('\0');
		if (!LoadFromExternalString((const char*)pBlob->GetData()))
			goto Cleanup;

		bSuccess = true;
		goto Cleanup;
		
		Cleanup:
			delete pBlob;
			return bSuccess;
	}

	bool Json::LoadFromBlob(gsBlob* pBlob)
	{
		if (!pBlob)
			return false;
		return LoadFromExternalString((const char*)pBlob->GetData());
	}

	Json* Json::GetNext()
	{
		return m_pNext;
	}

	Json* Json::GetChild()
	{
		return m_pChild;
	}

	bool Json::IsString() { return cJSON_IsString(m_pJson); }
	bool Json::IsNumber() { return cJSON_IsNumber(m_pJson); }
	bool Json::IsBool() { return cJSON_IsBool(m_pJson); }
	bool Json::IsObject() { return cJSON_IsObject(m_pJson); }
	bool Json::IsArray() { return cJSON_IsArray(m_pJson); }

	const char* Json::GetString() { return cJSON_GetStringValue(m_pJson); }
	double Json::GetNumber() { return cJSON_GetNumberValue(m_pJson); }
	bool Json::GetBool() { return cJSON_IsTrue(m_pJson); }

	Json* Json::GetField(const char* szName)
	{
		if (IsObject())
		{
			Json* pJson = m_pChild;
			while (pJson)
			{
				if (ExternalString::Equal(pJson->m_pJson->string, szName))
					return pJson;
				pJson = pJson->m_pNext;
			}
		}
		return NULL;
	}

	bool Json::RecursiveCreate()
	{
		if (!m_pJson)
			return false;

		if (m_pJson->next)
			m_pNext = new Json(m_pJson->next, this);
		
		if (m_pJson->child)
			m_pChild = new Json(m_pJson->child, NULL);

		return true;
	}
	
	void Json::Cleanup()
	{
		if (m_pOwnedJson)
			cJSON_Delete(m_pOwnedJson);
		m_pOwnedJson = NULL;
		m_pJson = NULL;
		if (m_pNext)
			delete m_pNext;
		m_pNext = NULL;
		m_pPrevious = NULL;

		if (m_pChild)
			delete m_pChild;
		m_pChild = NULL;
	}
}

File: Source/Json/CS/Json.cs
using System;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace NumbatLogic
{
	public class Json
	{
		private JsonNode m_pOwnedNode;
		private JsonNode m_pNode;

		private Json m_pPrevious;
		private Json m_pNext;
		private Json m_pChild;

		private string m_sName;

		public Json()
		{
			Cleanup();
		}

		private Json(JsonNode pNode, Json pPrevious, string sName)
		{
			m_pOwnedNode = null;
			m_pNode = pNode;
			m_pPrevious = pPrevious;
			m_pNext = null;
			m_pChild = null;
			m_sName = sName;

			RecursiveCreate();
		}

		public bool LoadFromExternalString(string sxIn)
		{
			Cleanup();

			if (sxIn == null)
				return false;

			try
			{
				JsonNode pNode = JsonNode.Parse(sxIn);
				if (pNode == null)
					return false;

				m_pOwnedNode = pNode;
				m_pNode = pNode;
				m_sName = null;

				RecursiveCreate();
				return true;
			}
			catch (Exception)
			{
				return false;
			}
		}

		public bool LoadFromPath(string sxPath)
		{
			bool bSuccess = false;
			gsBlob pBlob = new gsBlob();

			if (!pBlob.Load(sxPath))
				goto CleanupLabel;

			pBlob.SetOffset(0);
			if (!LoadFromBlob(pBlob))
				goto CleanupLabel;

			bSuccess = true;

		CleanupLabel:
			pBlob = null;
			return bSuccess;
		}

		public bool LoadFromBlob(gsBlob pBlob)
		{
			if (pBlob == null || pBlob.__pBuffer == null)
				return false;

			int nOffset = pBlob.GetOffset();
			int nSizeAvailable = pBlob.GetSize() - nOffset;
			if (nSizeAvailable <= 0)
				return false;

			string sxIn = Encoding.UTF8.GetString(pBlob.__pBuffer, nOffset, nSizeAvailable);
			return LoadFromExternalString(sxIn);
		}

		public Json GetNext()
		{
			return m_pNext;
		}

		public Json GetChild()
		{
			return m_pChild;
		}

		public bool IsString()
		{
			if (m_pNode is JsonValue jsonValue)
			{
				string s;
				return jsonValue.TryGetValue<string>(out s);
			}
			return false;
		}

		public bool IsNumber()
		{
			if (m_pNode is JsonValue jsonValue)
			{
				double d;
				return jsonValue.TryGetValue<double>(out d);
			}
			return false;
		}

		public bool IsBool()
		{
			if (m_pNode is JsonValue jsonValue)
			{
				bool b;
				return jsonValue.TryGetValue<bool>(out b);
			}
			return false;
		}

		public bool IsObject()
		{
			return m_pNode is JsonObject;
		}

		public bool IsArray()
		{
			return m_pNode is JsonArray;
		}

		public string GetString()
		{
			if (m_pNode is JsonValue jsonValue)
			{
				string s;
				if (jsonValue.TryGetValue<string>(out s))
					return s;
			}
			return null;
		}

		public double GetNumber()
		{
			if (m_pNode is JsonValue jsonValue)
			{
				double d;
				if (jsonValue.TryGetValue<double>(out d))
					return d;
			}
			return 0.0;
		}

		public bool GetBool()
		{
			if (m_pNode is JsonValue jsonValue)
			{
				bool b;
				if (jsonValue.TryGetValue<bool>(out b))
					return b;
			}
			return false;
		}

		public Json GetField(string sxName)
		{
			if (!IsObject() || sxName == null)
				return null;

			Json pJson = m_pChild;
			while (pJson != null)
			{
				if (ExternalString.Equal(pJson.m_sName, sxName))
					return pJson;
				pJson = pJson.m_pNext;
			}
			return null;
		}

		private bool RecursiveCreate()
		{
			if (m_pNode == null)
				return false;

			Json firstChild = null;
			Json previousChild = null;

			if (m_pNode is JsonObject jsonObject)
			{
				foreach (var kvp in jsonObject)
				{
					Json child = new Json(kvp.Value, previousChild, kvp.Key);
					if (firstChild == null)
						firstChild = child;

					if (previousChild != null)
						previousChild.m_pNext = child;

					previousChild = child;
				}
			}
			else if (m_pNode is JsonArray jsonArray)
			{
				for (int i = 0; i < jsonArray.Count; i++)
				{
					Json child = new Json(jsonArray[i], previousChild, null);
					if (firstChild == null)
						firstChild = child;

					if (previousChild != null)
						previousChild.m_pNext = child;

					previousChild = child;
				}
			}

			m_pChild = firstChild;
			return true;
		}

		private void Cleanup()
		{
			m_pOwnedNode = null;
			m_pNode = null;
			m_pPrevious = null;
			m_pNext = null;
			m_pChild = null;
			m_sName = null;
		}
	}
}

